main/streams/streams.c

```diff
@@ -1418,8 +1418,8 @@ PHPAPI zend_string *_php_stream_copy_to_mem(php_stream *src, size_t maxlen, int
 			ptr += ret;
 		}
 		if (len) {
-			*ptr = '\0';
-			ZSTR_LEN(result) = len;
+			result = zend_string_truncate(result, len, persistent);
+			ZSTR_VAL(result)[len] = '\0';
```

Comment:
You might want to use the same logic as in https://github.com/php/php-src/blob/master/main/streams/streams.c#L759 and only truncate if the "wastage" is large enough.

Reply:
@nikic thanks! I guess that the potential savings of the truncate are not worth the extra performance costs of the function and memory copy calls if the "waste" is not large enough?

Reply:
done!

Reply:
This dropped the `ZSTR_LEN(result) = len` assignment, which is still needed in case we don't truncate. I'm surprised this doesn't cause test failures.

Reply:
@nikic you're right, it didn't I'm afraid, any ideas for a test that maybe we could add?

Reply:
fixed in any case

Reply:
```
<?php
$f = fopen('php://memory', 'rw');
fwrite($f, str_repeat('X', 1000));
fseek($f, 0); 
var_dump(strlen(stream_get_contents($f, 1024)));
```

should show the issue.

Reply:
@nikic it did, thanks! I added the additional test just to be on the safe side